cmd/sops/common/common.go

```diff
@@ -136,3 +140,178 @@ func DefaultStoreForPath(path string) Store {
 	}
 	return &json.BinaryStore{}
 }
+
+const KMS_ENC_CTX_BUG_FIXED_VERSION = "3.3.0"
+
+func DetectKMSEncryptionContextBug(tree *sops.Tree) (bool, error) {
+	versionCheck, err := version.AIsNewerThanB(KMS_ENC_CTX_BUG_FIXED_VERSION, tree.Metadata.Version)
+	if err != nil {
+		return false, err
+	}
+
+	if versionCheck {
+		_, _, key := GetKMSKeyWithEncryptionCtx(tree)
+		if key != nil {
+			return true, nil
+		}
+	}
+
+	return false, nil
+}
+
+func GetKMSKeyWithEncryptionCtx(tree *sops.Tree) (kgndx int, keyndx int, key *kms.MasterKey) {
+	for i, kg := range tree.Metadata.KeyGroups {
+		for n, k := range kg {
+			kmsKey, ok := k.(*kms.MasterKey)
+			if ok {
+				if kmsKey.EncryptionContext != nil && len(kmsKey.EncryptionContext) >= 2 {
+					return i, n, kmsKey
+				}
+			}
+		}
+	}
+	return 0, 0, nil
+}
+
+type GenericDecryptOpts struct {
+	Cipher      sops.Cipher
+	InputStore  sops.Store
+	InputPath   string
+	IgnoreMAC   bool
+	KeyServices []keyservice.KeyServiceClient
+}
+
+func LoadEncryptedFileWithBugFixes(opts GenericDecryptOpts) (*sops.Tree, error) {
+	tree, err := LoadEncryptedFile(opts.InputStore, opts.InputPath)
+	if err != nil {
+		return nil, err
+	}
+
+	encCtxBug, err := DetectKMSEncryptionContextBug(tree)
+	if err != nil {
+		return nil, err
+	}
+	if encCtxBug {
+		tree, err = FixAWSKMSEncryptionContextBug(opts, tree)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	return tree, nil
+}
+
+func FixAWSKMSEncryptionContextBug(opts GenericDecryptOpts, tree *sops.Tree) (*sops.Tree, error) {
+	// TODO - Finalize messaging here.
+	message := "Up until version 3.3.0 of sops there was a bug surrounding the " +
+		"use of encryption context with AWS KMS."
+	fmt.Println(wordwrap.WrapString(message, 75))
+
+	if !terminal.IsTerminal(int(os.Stdout.Fd())) {
+		return nil, NewExitError("Need to run this from a tty to resolve this issue.", codes.TTYRequired)
+	}
+
+	var response string
+	for response != "y" && response != "n" {
+		fmt.Println("Would you like sops to automatically fix this issue? (y/n): ")
+		_, err := fmt.Scanln(&response)
+		if err != nil {
+			return nil, err
+		}
+	}
+	if response == "n" {
+		return nil, fmt.Errorf("Exiting. User responded no.")
+	}
+
+	dataKey := []byte{}
+	success := false
+	// If there is another key, then we should be able to just decrypt
+	// without having to try different variations of the encryption context.
+	dataKey, err := DecryptTree(DecryptTreeOpts{
+		Cipher:      opts.Cipher,
+		IgnoreMac:   opts.IgnoreMAC,
+		Tree:        tree,
+		KeyServices: opts.KeyServices,
+	})
+	if err != nil {
+		kgndx, kndx, originalKey := GetKMSKeyWithEncryptionCtx(tree)
+
+		keyToEdit := *originalKey
+
+		encCtxVals := []string{}
```

Comment:
Nit: could make this a `map[string]struct{}` since we don't need to handle duplicate values